### ISSUE TITLE
Wrong `Lines to Cover` metric when using parameters with default values

### ISSUE THREAD

**Issue by lgregoire-yseop** [SonarSource/SonarTS#831]
## TL;DR

I found a weird behavior with the count of `Lines to Cover` when using default values for methods' parameters in TypeScript.

## Reproducer
*index.ts*
```ts
export class Foo {
    public bar(
        param1: string = 'PARAM1',
        param2: string = 'PARAM2',
        param3: string = 'PARAM3',
        param4: string = 'PARAM4',
    ) {
        return `${param1} - ${param2} - ${param3} - ${param4}`;
    }
}
```

*index.test.ts*
```ts
import { Foo } from './';

describe('Foo', () => {
    describe('bar', () => {
        it('should return a concatenation of all parameters', () => {
            expect(new Foo().bar('p1', 'p2', 'p3', 'p4')).toBe('p1 - p2 - p3 - p4');
        });
    });
});

```

At this point, we agree that none of the default parameters of `bar` are covered by tests.
So `jest` returns this following coverage report:

![image](https://user-images.githubusercontent.com/4543679/53160726-d1bbfc00-35c8-11e9-9b1c-8b7896b59f5a.png)

### Analysis

Now, I run `sonar-scanner -Dsonar.branch.base=master -Dsonar.branch.name=test` to get a report on my new code:

![image](https://user-images.githubusercontent.com/4543679/53161692-be119500-35ca-11e9-9420-73628ff9b130.png)

And here is the **weird behavior**:
- The previous coverage on this file was `100%`
- I covered only 1 of 5 new lines so the new code coverage should be `20%` and `33.3%` after merge
- But Sonar ends with: 
  - `Lines to Cover` = `1` instead of `5`
  - `Uncovered Lines` = `0` instead of `4`
  - `Coverage on New Code` = `100%` instead of `20%`

Finally, it seems that parameters with default values are not counted as lines.

## Informations
- **SonarTS version:** 1.9.0.3766
- **Node.js version:** 6.8.0
- **TypeScript version:** 3.3.3

Repo: https://github.com/lgregoire-yseop/SonarTS-coverage-test
SonarCloud project: https://sonarcloud.io/dashboard?id=SonarTS-coverage-test

**Comment by vilchik-elena** [comment]
Can you show coverage report you provide to Sonar?

**Comment by lgregoire-yseop** [comment]
Yes, here is my `lcov.info`:
```
TN:
SF:/home/lgregoire/Documents/workspace/SonarTS-coverage-test/src/index.ts
FN:2,(anonymous_0)
FNF:1
FNH:1
FNDA:1,(anonymous_0)
DA:1,1
DA:8,1
LF:2
LH:2
BRDA:3,0,0,0
BRDA:4,1,0,0
BRDA:5,2,0,0
BRDA:6,3,0,0
BRF:4
BRH:0
end_of_record
```

**Comment by vilchik-elena** [comment]
Sonar does not guess which lines need to be covered when there is coverage report for this file (in other words when there is no info, to have meaningful not-covered lines we compute lines to cover).
When there is a report, we entirely rely on it. 

So in your case I would expect to see in your report
```
DA: 3,0
DA: 4,0
DA: 5,0
DA: 6,0
```
Which means that these lines should be covered but they are not. May be there is some jest property which allow you to get this.

Then about coverage on new code (100% in your case), I think condition coverage is not considered there. 

So at the end it's pretty a corner case, you have conditional code on the line, but it might be not executed. Usually in case of conditional code (e.g. `if (cond) {...`) you will execute this line whatever the result.

Please check with your tool if it's possible to generate data so that's correctly displayed in Sonar. If not, let's see how we can interpret current report (note that we don't match to syntax at this stage, we have nothing but coverage report).

Hope it helps

**Comment by lgregoire-yseop** [comment]
Ok I got the point.
In fact, I thought it was a weird behavior of Sonar because the Istanbul text reporter displays the uncovered branches in the `Uncovered Lines` column when lines coverage = `100%`.

![image](https://user-images.githubusercontent.com/4543679/53232205-e6f95f00-368a-11e9-8769-09ba62e303cf.png)

But when there is a real uncovered line (eg. `DA:9,0`), the uncovered branches disappear:
![image](https://user-images.githubusercontent.com/4543679/53232238-f5477b00-368a-11e9-8fe5-e7f530bc81df.png)

Thanks for your answers @vilchik-elena :+1:
